### PR TITLE
Support broadcasting in `simulate` and `preprocess`

### PR DESCRIPTION
**Pull request by eddddddy**
**To be merged after https://github.com/PennyLaneAI/pennylane/pull/4238**

**Context:**
Part of the development of the new qubit device in PennyLane.

**Description of the Change:**
Support broadcasting in the `simulate` function and in the preprocessing functions in `devices/qubit/preprocess.py`. Built-in PennyLane broadcasting will now used when executing qscripts in the new device, unless the execution config specifies `gradient_method="adjoint"`, in which case `qml.transforms.broadcast_expand` will still be used to split executions.

### CODE REVIEW

File: pennylane/devices/qubit/preprocess.py [hunk 1]
```diff
@@ -258,15 +259,18 @@ def batch_transform(
...
         the sequence of circuits to be executed, and a post-processing function
         to be applied to the list of evaluated circuit results.
     """
-    # Check whether the circuit was broadcasted
-    if circuit.batch_size is None:
-        # If the circuit wasn't broadcasted, no action required
+    # Check whether the circuit was broadcasted or if the diff method is anything other than adjoint
+    if circuit.batch_size is None or execution_config.gradient_method != "adjoint":
+        # If the circuit wasn't broadcasted, or if built-in PennyLane broadcasting
+        # can be used, then no action required
```
Comment (josh146) [inline pennylane/devices/qubit/preprocess.py:273]: Hey @eddddddy! I have a random fly-by question. Are these preprocessing functions general ones, meant to be used for any qubit device, or are they implementation details of `default.qubit2`, and not meant to be used for other devices?

Comment (eddddddy) [reply pennylane/devices/qubit/preprocess.py:273]: As far as I understand, these are only to be used by `default.qubit2`.

File: pennylane/devices/qubit/simulate.py [hunk 2]
```diff
@@ -57,16 +57,26 @@ def simulate(circuit: qml.tape.QuantumScript, rng=None, debugger=None) -> Result
 
     state = create_initial_state(circuit.wires, circuit._prep[0] if circuit._prep else None)
 
+    # initial state is batched only if the state preparation (if it exists) is batched
+    is_state_batched = False
```
Comment (mudit2812) [inline pennylane/devices/qubit/simulate.py:61]: Non-blocking comment but I'm not personally a big fan of naming the variable `is_state_batched`. I think a slightly more concise name would be `batched_state`. My personal opinion is to usually avoid using more than one underscore in a variable name. Same thing for the function arguments below.

**Review by mudit2812 — APPROVED** [review]
Looks good to me! I left a non-blocking suggestion.

File: pennylane/devices/qubit/preprocess.py (hunk 1 above)
Comment (albi3ro) [reply pennylane/devices/qubit/preprocess.py:273]: Now that the new transforms stuff is starting to get merged, we are going to make ransforms for all these steps make it easy to pick and choose what steps a particular device wants.

File: pennylane/devices/qubit/simulate.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/devices/qubit/simulate.py:61]: Just linking back to this discussion: https://github.com/PennyLaneAI/pennylane/pull/3852#discussion_r1124951529 .

This is all fairly internal, so I think we'd be fine to change this name around freely. At least for now.  But I think this is a different discussion we can kick down the road.

Comment (timmysilv) [reply pennylane/devices/qubit/simulate.py:61]: that kwarg was part of 0.30 in `apply_operation.py` and I don't really wanna deprecate it. It wasn't in any public docs... but still

PS: check the user who [committed the variable name](https://github.com/PennyLaneAI/pennylane/commit/1920f9da1939555e99eb29232be6f6e68c9e9ad6) 😆

Comment (timmysilv) [reply pennylane/devices/qubit/simulate.py:61]: oops too slow

**Review by timmysilv — APPROVED** [review]

File: pennylane/devices/qubit/simulate.py (hunk 2 above)
Comment (eddddddy) [reply pennylane/devices/qubit/simulate.py:61]: This is not super high priority, and according to the linked discussion the name `is_state_batched` satisfies most people. So I'd like to keep it as is at least for now.

**Comment by eddddddy** [comment]
[sc-40107]

**Pull request merged**